### ISSUE TITLE
BC13.0 has critical limitations in use of Enums, e.g. no "Quoted Names", no InitValue - but VS Code reports no problems - failure only occurs on publish

### ISSUE THREAD

**Issue by dzzzb** [microsoft/AL#7060]
The title says it all:

Here is VS Code's view of a project using runtime 2.0 in `app.json` but also using `Enum`s. See how it conflicts with the server's view!

![image](https://user-images.githubusercontent.com/63455896/170738792-0a968546-3aa3-4d1c-b274-f0de914f1c53.png)

[Apparently](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/enum/enum-data-type) `Enum`s are only valid from runtime version 4+.

Thus the VS Code AL extension should let us know about this early, rather than indicating everything is OK and then erroring only when we try to publish to a server with incompatible runtime.

Thanks!

```none
Name: AL Language
Id: ms-dynamics-smb.al
Description: AL development tools for Dynamics 365 Business Central
Version: 9.0.615906
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-dynamics-smb.al
```

**Comment by blrobl** [comment]
Hi, this does not seem to be an error related to enums not being valid, since they are available since runtime version 2.0 (documentation refers to enums having been modified in version 4.0, not introduced). It looks like your local app is not synced with the server and that's what could be causing the error - could you try to download the symbols from the server and see if the issue persists? Thanks.

**Comment by dnpb** [comment]
> It looks like your local app is not synced with the server and that's what could be causing the error

No, there is nothing to sync; these enums are declared and used as table fields within the same app. They aren't from other apps. So downloading symbols is not applicable.

Toy projects do not readily show the problem. I'll have to work on a reduced repro from the actual app that does, to figure out where it goes wrong.

***

> (documentation refers to enums having been modified in version 4.0, not introduced)

That's not ideal. Shouldn't it say when things were first made available, instead of being ambiguous? If you have access to that info, can users/readers have it too? Thanks!

**Comment by dnpb** [comment]
So far I can't replicate all the publish-time problems, but I _can_ this one:

```al
enum 50100 TestEnum
{
    value(1; TestValue1) { }
}

table 50100 TestTable
{
    fields
    {
        field(50100; TestEnumField; Enum TestEnum)
        {
            InitValue = TestValue1;

            trigger OnValidate()
            begin
                if Rec.TestEnumField <> Rec.TestEnumField::TestValue1 then
                    Error('balp.');
            end;
        }
    }
}
```

VS Code reports no Problems.

But on Publishing to a BC13 server and building there:
```none
[2022-09-09 14:35:10.67] The request for path /NAV/dev/apps?tenant=default&SchemaUpdateMode=synchronize&DependencyPublishingOption=default failed with code UnprocessableEntity. Reason: Extension compilation failed
Table.al(7,25): error AL0294: The type of property value TestValue1 does not match the field’s type.
Table.al(7,13): error AL0167: The property 'InitValue' can only be set if the property 'Type' is set with any of the values of :'BigInteger,Boolean,Code,Date,DateFormula,DateTime,Decimal,Integer,Option,Text,Time,TextConst'
```

I will update as I manage to replicate more of my original errors, which were failure of the BC13 server to recognise members referenced by `Rec.Field::Syntax` or `pParameter::FieldSyntax`.

**Comment by dnpb** [comment]
Here's the others replicated. The problem seems to be simply having `"Quotes" in value names. So BC13 just does not support that, or...?!?

```al
enum 50100 TestEnum
{
    value(1; "Test Value 1") { }
}

table 50100 TestTable
{
    fields
    {
        field(50100; TestEnumField; Enum TestEnum)
        {
            InitValue = "Test Value 1";

            trigger OnValidate()
            begin
                if Rec.TestEnumField <> Rec.TestEnumField::"Test Value 1" then
                    Error('balp.');
            end;
        }
    }
}

codeunit 50100 Test
{
    procedure TestParam(e: Enum TestEnum)
    begin
        if e <> e::"Test Value 1" then
            Error('wut');
    end;
}
```

giving:

```none
[2022-09-09 14:37:48.17] Publishing AL application using launch configuration 'bc1300-20220523'.
[2022-09-09 14:37:48.56] Publishing package to tenant 'default'
[2022-09-09 14:37:48.56] Targeting server 'http://bc1300-20220523', server instance 'NAV' and tenant 'default'.
[2022-09-09 14:37:48.57] Using user name and password authentication. User name used is: 'admin'.
[2022-09-09 14:37:48.58] Sending request to http://bc1300-20220523:7049/NAV/dev/apps?tenant=default&SchemaUpdateMode=synchronize&DependencyPublishingOption=default
[2022-09-09 14:37:48.87] The request for path /NAV/dev/apps?tenant=default&SchemaUpdateMode=synchronize&DependencyPublishingOption=default failed with code UnprocessableEntity. Reason: Extension compilation failed
Enum.al(12,25): error AL0294: The type of property value "Test Value 1" does not match the field’s type.
Enum.al(12,13): error AL0167: The property 'InitValue' can only be set if the property 'Type' is set with any of the values of :'BigInteger,Boolean,Code,Date,DateFormula,DateTime,Decimal,Integer,Option,Text,Time,TextConst'
Enum.al(27,20): error AL0132: 'e' does not contain a definition for 'Test Value 1'
Enum.al(16,60): error AL0132: 'TestEnumField' does not contain a definition for 'Test Value 1'
```

So it seems that runtime 2.0 / BC13 are critically limited when it comes to what we can actually do with Enums... then, as I originally said, VS Code should be aware of these limitations and report them as Problems instead of deferring to publish time.

**Comment by dnpb** [comment]
related: https://github.com/microsoft/AL/issues/3566

**Comment by blrobl** [comment]
Thanks for the quick and helpful response! I have checked and this is indeed due to two bugs on our side; quoted enum values are not handled correctly and some properties are not supported for Enums. 
Both were fixed from BC 13.1, so I can suggest you to upgrade to a higher version. Unfortunately these will not be fixed for version 13.0, since it is [no longer supported](https://docs.microsoft.com/en-us/lifecycle/products/dynamics-365-business-central-onpremises-modern-policy).

**Comment by dnpb** [comment]
Thanks! I don't expect them to be fixed 'in' version 13.0, but it would be great if the AL language extension could be made aware of these big limitations, and warn/error to users who try using them when targeting BC13.0 - is that not possible?

**Comment by blrobl** [comment]
We have decided that we will not be making any changes just for BC13.0 unfortunately, since it's long out of support.
